lib/async_storage.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+const { createHook, executionAsyncId } = require('async_hooks');
+const { WeakReference } = internalBinding('util');
+const { emitExperimentalWarning } = require('internal/util');
+
+emitExperimentalWarning('async_storage');
+// id => WeakReference<AsyncResource>
+const idResourcePairing = new Map();
+const storages = new Set();
+createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    if (triggerResourceWeak) { // it has a parent, let it be known
+      storages.forEach((storage) => {
+        storage._propagate(resource, triggerResourceWeak.get());
+      });
+    }
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+})
+  .enable();
```

Comment:
I don’t think this is a good idea. We should activate this manually, not on require.

Reply:
Ack, will do